src/moscot/problems/base/_utils.py

```diff
@@ -228,3 +236,242 @@ def _filter_kwargs(*funcs: Callable[..., Any], **kwargs: Any) -> Dict[str, Any]:
+
+    n = X.shape[1]  # genes x cells
+    ql = 1 - confidence_level - (1 - confidence_level) / 2.0
+    qh = confidence_level + (1 - confidence_level) / 2.0
+
+    if issparse(X) and not isspmatrix_csr(X):
+        X = csr_matrix(X)
+
+    corr = _mat_mat_corr_sparse(X, Y) if issparse(X) else _mat_mat_corr_dense(X, Y)
+
+    if method == CorrTestMethod.FISCHER:
+        # see: https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#Using_the_Fisher_transformation
+        mean, se = np.arctanh(corr), 1.0 / np.sqrt(n - 3)
+        z_score = (np.arctanh(corr) - np.arctanh(0)) * np.sqrt(n - 3)
+
+        z = norm.ppf(qh)
+        corr_ci_low = np.tanh(mean - z * se)
+        corr_ci_high = np.tanh(mean + z * se)
+        pvals = 2 * norm.cdf(-np.abs(z_score))
+
+    elif method == CorrTestMethod.PERM_TEST:
+        if not isinstance(n_perms, int):
+            raise TypeError(f"Expected `n_perms` to be an integer, found `{type(n_perms).__name__}`.")
+        if n_perms <= 0:
+            raise ValueError(f"Expcted `n_perms` to be positive, found `{n_perms}`.")
+
+        pvals, corr_ci_low, corr_ci_high = parallelize(
+            _perm_test,  # type:ignore[arg-type]
+            np.arange(n_perms),
+            as_array=False,
+            unit="permutation",
+            extractor=perm_test_extractor,
+            **kwargs,
+        )(corr, X, Y, seed=seed)
+
+    else:
+        raise NotImplementedError(method)
+
+    return corr, pvals, corr_ci_low, corr_ci_high
+
+
+def _mat_mat_corr_sparse(
+    X: csr_matrix,
+    Y: ArrayLike,
+) -> ArrayLike:
+    n = X.shape[1]
+    X_bar = np.reshape(np.array(X.mean(axis=1)), (-1, 1))
+    X_std = np.reshape(np.sqrt(np.array(X.power(2).mean(axis=1)) - (X_bar**2)), (-1, 1))
+
+    y_bar = np.reshape(np.mean(Y, axis=0), (1, -1))
+    y_std = np.reshape(np.std(Y, axis=0), (1, -1))
+
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", RuntimeWarning)
+        return (X @ Y - (n * X_bar * y_bar)) / ((n - 1) * X_std * y_std)
+
+
+def _mat_mat_corr_dense(X: ArrayLike, Y: ArrayLike) -> ArrayLike:
+    from moscot._utils import np_std, np_mean
+
+    n = X.shape[1]
+
+    X_bar = np.reshape(np_mean(X, axis=1), (-1, 1))
+    X_std = np.reshape(np_std(X, axis=1), (-1, 1))
+
+    y_bar = np.reshape(np_mean(Y, axis=0), (1, -1))
+    y_std = np.reshape(np_std(Y, axis=0), (1, -1))
+
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", RuntimeWarning)
+        return (X @ Y - (n * X_bar * y_bar)) / ((n - 1) * X_std * y_std)
+
+
+def _perm_test(
+    ixs: ArrayLike,
+    corr: ArrayLike,
+    X: Union[ArrayLike, spmatrix],
+    Y: ArrayLike,
+    seed: Optional[int] = None,
+    queue=None,
+) -> Tuple[ArrayLike, ArrayLike]:
+    rs = np.random.RandomState(None if seed is None else seed + ixs[0])
+    cell_ixs = np.arange(X.shape[1])
+    pvals = np.zeros_like(corr, dtype=np.float64)
```

Comment:
Might need some cleaning-up, as far as I remember, `np.float64` alias has been deprecated.

Reply:
just checked and didn't find any indication it was deprecated, see e.g. https://numpy.org/doc/stable/user/basics.types.html

Reply:
You're right, my mistake, these are the [deprecated aliases](https://numpy.org/devdocs/release/1.20.0-notes.html#using-the-aliases-of-builtin-types-like-np-int-is-deprecated) (e.g., `np.float`).